src/core/quantum_gate.jl

```diff
@@ -147,6 +161,52 @@ function apply_gate!(state::Ket, gate::Gate)
     Snowflake.apply_gate_without_ket_size_check!(state, gate, Int(qubit_count))
 end
 
+function apply_gate!(state::Ket, gate::AbstractGate)
+    qubit_count = log2(length(state))
+    if mod(qubit_count, 1) != 0
+        throw(DomainError(qubit_count,
+            "Ket does not correspond to an integer number of qubits"))
+    end
+    if gate.target>qubit_count #TODO: implement for multi-target gates
+        throw(DomainError(gate.target,
+            "not enough qubits in the Ket for the Gate"))
+    end
+
+    connected_qubits=get_connected_qubits(gate)
+
+    type_in_ket=eltype(state.data)
+
+    operator=get_operator(gate,type_in_ket)
+
+    apply_operator!(state,operator,connected_qubits,Int(qubit_count))
+
+end
+
+get_connected_qubits(gate::AbstractGate)=
+    throw(NotImplementedError(:get_connected_qubits, gate))
+
+
+function apply_operator!(state::Ket,operator::DiagonalOperator,connected_qubit::Int,qubit_count::Int)
+
+    dim=2^qubit_count
+    target_qubit_index=qubit_count-connected_qubit # indexing of targets in qulacs starts at 0
+    
+    diagonal_in_matrix=operator.data
+
```

Comment:
This implementation assumes a single qubit operator. We need to generalize it or mark it as a specialization.

Comparatively, what performance do we get when implementing the diagonal operator using `@simd` in the loop?